// @sap/sbf/lib/credential-providers/SbssProvider.js
'use strict';

const sbssLib = require('@sap/sbss');

class SbssProvider {
  constructor(credentials) {
    this._sbss = sbssLib(credentials);
  }

  provision(req, cb) { cb(); }

  deprovision(req, cb) {
    this._sbss.deleteAllInstanceCredentials(req.brokerContext.instance_id, err => cb(err));
  }

  bind(req, cb) {
    const bindingInfo = {
      instanceId: req.brokerContext.instance_id,
      bindingId: req.brokerContext.binding_id,
      serviceId: req.brokerContext.service_id,
      planId: req.brokerContext.plan_id
    };

    const appGuid = (req.brokerContext.bind_resource && req.brokerContext.bind_resource.app_guid) || req.brokerContext.app_guid;
    if (appGuid) {
      bindingInfo.appGuid = appGuid;
    }

    const subaccountId = req.brokerContext.context && req.brokerContext.context.subaccount_id;
    if (subaccountId) {
      bindingInfo.subaccountId = subaccountId;
    }

    this._sbss.createCredentials(bindingInfo, cb);
  }

  unbind(req, cb) {
    this._sbss.deleteCredentials(req.brokerContext.instance_id, req.brokerContext.binding_id, err => cb(err));
  }
}

module.exports = SbssProvider;
